# SPDX-License-Identifier: Apache-2.0
From 1ceb9972eb460090fc8cc8524d9f036f562cb552 Mon Sep 17 00:00:00 2001
From: Andrew Bogott <abogott@wikimedia.org>
Date: Fri, 01 Sep 2023 18:14:43 -0500
Subject: [PATCH] config loader: Prefer cli/env over clouds.yaml args for some args

If a user specifies a project or domain on the cli, that should take
precedence over the value set in clouds.yaml. This fixes enables workflows
that rely on domain- or cloud-wide credentials in clouds.yaml.

The cli args that are reprioritized in this patch are:

 domain-id
 domain-name
 user-domain-id
 user-domain-name
 project-domain-id
 project-domain-name
 auth-token
 project-id
 tenant-id
 project-name
 tenant-name

Story: 2010784
Change-Id: I45e7cff6579e6686d790bd3bb3e3ab9955885a64
---
diff --git a/openstack/config/loader.py b/openstack/config/loader.py
index 1f330e8..2954e16 100644
--- a/openstack/config/loader.py
+++ b/openstack/config/loader.py
@@ -777,12 +777,15 @@
         for target_key, possible_values in mappings.items():
             target = None
             for key in possible_values:
-                if key in cloud:
-                    target = str(cloud[key])
-                    del cloud[key]
+                # Prefer values from the 'auth' section
+                # as they may contain cli or environment overrides.
+                # See story 2010784 for context.
                 if key in cloud['auth']:
                     target = str(cloud['auth'][key])
                     del cloud['auth'][key]
+                if key in cloud:
+                    target = str(cloud[key])
+                    del cloud[key]
             if target:
                 cloud['auth'][target_key] = target
         return cloud